src/shared/sam/debugger/awsSamDebugger.ts

```diff
@@ -236,3 +438,60 @@ function codeDebugConfigValidation(
 
     return { isValid: true }
 }
+
+function getCodeRoot(
+        folder: vscode.WorkspaceFolder | undefined,
+        config: AwsSamDebuggerConfiguration
+        ): string | undefined {
+    switch (config.invokeTarget.target) {
+        case 'code': {
+            const codeInvoke = (config.invokeTarget as CodeTargetProperties)
+            return tryGetAbsolutePath(folder, codeInvoke.projectRoot)
+        }
+        case 'template': {
+            const templateInvoke = (config.invokeTarget as TemplateTargetProperties)
+            const templateResource = getTemplateResource(config)
+            if (!templateResource?.Properties) {
+                return undefined
+            }
+            const templateDir = path.dirname(templateInvoke.samTemplatePath)
+            return path.resolve(templateDir ?? '', templateResource?.Properties?.CodeUri)
+        }
+        default: {
+            throw Error('invalid invokeTarget')  // Must not happen.
+        }
+    }
+}
+
+function getHandlerName(config: AwsSamDebuggerConfiguration): string {
+    switch (config.invokeTarget.target) {
+        case 'code': {
+            const codeInvoke = (config.invokeTarget as CodeTargetProperties)
+            return codeInvoke.lambdaHandler
+        }
+        case 'template': {
+            const templateResource = getTemplateResource(config)
+            return templateResource?.Properties?.Handler!!
```

Comment:
Would this potentially return `undefined`? If so, should we throw an error/replace with a blank string?

Reply:
shouldn't `!!` already cause an error in that case?

Reply:
I don't think `!!` means anything different than `!` (at least, I can't find anything documenting it).

It appears that `!` is only a compiler fix (so yes, this will throw an error): https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator ; would we want to handle said error more gracefully?

Reply:
error should never happen, this is just asserting an invariant. If the error does happen the problem should be fixed earlier in the codepath.

btw we may need to compile with `--strictNullChecks`